src/classic/clvm_tools/stages/stage_2/module.rs

```diff
@@ -385,26 +397,110 @@ fn compile_mod_stage_1(
+
+                    for (name, delayed_body) in delayed_constant_defs.iter() {
+                        let main_list =
+                            enlist(
+                                allocator,
+                                &[allocator.null(), *delayed_body]
+                            )?;
+
+                        result_collection.functions.insert(
+                            main_name_vec.clone(), main_list
+                        );
+
+                        let used_in_this_constant = build_used_constants_names(
                             allocator,
-                            &[main_local_arguments, uncompiled_main]
+                            &result_collection.functions,
+                            &delayed_constants,
+                            &result_collection.macros
+                        )?;
+
+                        let uses_other_constants = used_in_this_constant.iter().
+                            any(|u| delayed_constants.contains_key(u));
+
+                        if uses_other_constants {
+                            continue;
+                        }
+
+                        processed = true;
+
+                        let compiled =
+                            finish_compile_from_collection(
+                                allocator,
+                                macro_lookup,
+                                run_program.clone(),
+                                &result_collection,
+                                produce_extra_info
+                            )?;
+
+                        let compilation_result =
+                            run_program.run_program(
+                                allocator,
+                                compiled,
+                                allocator.null(),
+                                None
+                            )?;
+
+                        let result =
+                            run_program.run_program(
+                                allocator,
+                                compilation_result.1,
+                                allocator.null(),
+                                None
+                            )?;
+
+                        delayed_constants.remove(name);
+                        result_collection.constants.insert(
+                            name.to_vec(), quote(allocator, result.1)?
                         );
+                    }
 
-                    let _ = functions.insert(MAIN_NAME.as_bytes().to_vec(), main_list);
-                    Ok(CollectionResult {
-                        functions,
-                        constants,
-                        macros
-                    })
+                    if !processed {
+                        return Err(EvalErr(allocator.null(), "got stuck untangling defconst dependencies".to_string()));
+                    }
                 }
+
+                let uncompiled_main = alist[alist.len() - 1];
```

Comment:
would `alist.last();` work here?

Reply:
Probably.  I should get in that habit.  I need to come up with a good pattern for unexpected error in those cases, as exposing the Option gives the possibility of catching the err and one likely should.

Reply:
ok nice error bubbling.